libraries/extractor/src/main/java/androidx/media3/extractor/ts/H265Reader.java

```diff
@@ -246,216 +246,29 @@ private static Format parseMediaFormat(
-    if (bitArray.readBit()) { // pcm_enabled_flag
-      // pcm_sample_bit_depth_luma_minus1 (4), pcm_sample_bit_depth_chroma_minus1 (4)
-      bitArray.skipBits(8);
-      bitArray.readUnsignedExpGolombCodedInt(); // log2_min_pcm_luma_coding_block_size_minus3
-      bitArray.readUnsignedExpGolombCodedInt(); // log2_diff_max_min_pcm_luma_coding_block_size
-      bitArray.skipBit(); // pcm_loop_filter_disabled_flag
-    }
-    // Skips all short term reference picture sets.
-    skipShortTermRefPicSets(bitArray);
-    if (bitArray.readBit()) { // long_term_ref_pics_present_flag
-      // num_long_term_ref_pics_sps
-      for (int i = 0; i < bitArray.readUnsignedExpGolombCodedInt(); i++) {
-        int ltRefPicPocLsbSpsLength = log2MaxPicOrderCntLsbMinus4 + 4;
-        // lt_ref_pic_poc_lsb_sps[i], used_by_curr_pic_lt_sps_flag[i]
-        bitArray.skipBits(ltRefPicPocLsbSpsLength + 1);
-      }
-    }
-    bitArray.skipBits(2); // sps_temporal_mvp_enabled_flag, strong_intra_smoothing_enabled_flag
-    float pixelWidthHeightRatio = 1;
-    if (bitArray.readBit()) { // vui_parameters_present_flag
-      if (bitArray.readBit()) { // aspect_ratio_info_present_flag
-        int aspectRatioIdc = bitArray.readBits(8);
-        if (aspectRatioIdc == NalUnitUtil.EXTENDED_SAR) {
-          int sarWidth = bitArray.readBits(16);
-          int sarHeight = bitArray.readBits(16);
-          if (sarWidth != 0 && sarHeight != 0) {
-            pixelWidthHeightRatio = (float) sarWidth / sarHeight;
-          }
-        } else if (aspectRatioIdc < NalUnitUtil.ASPECT_RATIO_IDC_VALUES.length) {
-          pixelWidthHeightRatio = NalUnitUtil.ASPECT_RATIO_IDC_VALUES[aspectRatioIdc];
-        } else {
-          Log.w(TAG, "Unexpected aspect_ratio_idc value: " + aspectRatioIdc);
-        }
-      }
-      if (bitArray.readBit()) { // overscan_info_present_flag
-        bitArray.skipBit(); // overscan_appropriate_flag
-      }
-      if (bitArray.readBit()) { // video_signal_type_present_flag
-        bitArray.skipBits(4); // video_format, video_full_range_flag
-        if (bitArray.readBit()) { // colour_description_present_flag
-          // colour_primaries, transfer_characteristics, matrix_coeffs
-          bitArray.skipBits(24);
-        }
-      }
-      if (bitArray.readBit()) { // chroma_loc_info_present_flag
-        bitArray.readUnsignedExpGolombCodedInt(); // chroma_sample_loc_type_top_field
-        bitArray.readUnsignedExpGolombCodedInt(); // chroma_sample_loc_type_bottom_field
-      }
-      bitArray.skipBit(); // neutral_chroma_indication_flag
-      if (bitArray.readBit()) { // field_seq_flag
-        // field_seq_flag equal to 1 indicates that the coded video sequence conveys pictures that
-        // represent fields, which means that frame height is double the picture height.
-        picHeightInLumaSamples *= 2;
-      }
-    }
+    NalUnitUtil.H265SpsData spsData =
+        NalUnitUtil.parseH265SpsNalUnitPayload(sps.nalData, 40/8, sps.nalLength);
```

Comment:
I'm not sure why we need this `40/8` offset. There were 40 bits skipped in the original code.

```
bitArray.skipBits(40 + 4); // NAL header, sps_video_parameter_set_id
````

Reply:
Maybe Start code (4Byte) + NAL header(1Byte) ?

Reply:
yeah, i didn't quite get this bit either at first - I think this is what's going on:

The NAL header is 2 bytes, right? According to 7.3.1.2

The other 3 bytes are a NAL Unit start code (`0x0000001`) that is synthesised by `NalUnitTargetBuffer` in its constructor:

https://github.com/androidx/media/blob/3c01488f8d478caaa54038431ee6339444ea6e04/libraries/extractor/src/main/java/androidx/media3/extractor/ts/NalUnitTargetBuffer.java#L38-L40

---

Given this, I think we can actually tweak this even further to use `NalUnitUtil.parseH265SpsNalUnit` and pass `3` in (to just skip the start code). This has the very minor advantage of being consistent with the other callsites of `NalUnitUtil` for parsing H265 SPS:

https://github.com/androidx/media/blob/3c01488f8d478caaa54038431ee6339444ea6e04/libraries/extractor/src/main/java/androidx/media3/extractor/HevcConfig.java#L86-L88

https://github.com/androidx/media/blob/3c01488f8d478caaa54038431ee6339444ea6e04/libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/RtspMediaTrack.java#L397-L399